xml serializer sec hardening

cquirosj opened the pull request:
adding XML serializer change and tests for security hardening

timbussmann commented on src/NServiceBus.Core.Tests/Serializers/XML/SerializerTests.cs:
```diff
@@ -1227,10 +1227,78 @@ public void Object_property_with_primitive_or_struct_value_should_serialize_corr
+        static XmlMessageSerializer SerializerInstanceWithZeroMessageTypes() => SerializerFactory.Create();
+
+        [Test]
+        public void Should_throw_exception_when_deserializing_payloads_with_nsb_types()
```
can we add some actual reference to `ReplyOptions` so that this test indicates an issue in case we remove the used public type for some reason while not adjusting the test?

danielmarbach commented on src/NServiceBus.Core.Tests/Serializers/XML/SerializerTests.cs:
```diff
@@ -1227,10 +1227,78 @@ public void Object_property_with_primitive_or_struct_value_should_serialize_corr
+
+            using (var stream = new MemoryStream())
+            {
+                var streamWriter = new StreamWriter(stream);
```
Isn't streamwriter disposable too? Not that it really matters for the test but then you could it move on the same level like the MemStream. Also can we start using the new using syntax?

timbussmann replied:
I don't think we can use the new using syntax (c# 8) with .net472 as a target?

danielmarbach replied:
Using is one of the features that works https://stackoverflow.com/questions/56651472/does-c-sharp-8-support-the-net-framework

danielmarbach commented on src/NServiceBus.Core.Tests/Serializers/XML/SerializerTests.cs:
```diff
@@ -1227,10 +1227,78 @@ public void Object_property_with_primitive_or_struct_value_should_serialize_corr
+        {
+            var xml = @"<?xml version=""1.0""?><ArrayList xmlns:xsi=""http://www.w3.org/2001/XMLSchema-instance"" xmlns:xsd=""http://www.w3.org/2001/XMLSchema"" xmlns=""http://tempuri.net/System.Collections""></ArrayList>";
+
+            using (var stream = new MemoryStream())
+            {
+                var streamWriter = new StreamWriter(stream);
+                streamWriter.Write(xml);
+                streamWriter.Flush();
+                stream.Position = 0;
```
Seems to be a common setup that could be provided with some helper. Normally, I go with don't abstract before you have used it three times as my own personal rule of thumb which we would have crossed here

cquirosj commented:
@timbussmann  and @danielmarbach I addressed your comments:

- Moving one test to another type in order to reuse nested types from the same .cs file
- Ensuring dispose is called for StreamWriter
- Removing blank lines
- Removing duplication
- Adding an actual reference to ReplyOptions

Could you take another look? Thanks

danielmarbach commented on src/NServiceBus.Core.Tests/Serializers/XML/SerializerTests.cs:
```diff
@@ -1227,6 +1227,49 @@ public void Object_property_with_primitive_or_struct_value_should_serialize_corr
+
+        static XmlMessageSerializer SerializerInstanceWithZeroMessageTypes() => SerializerFactory.Create();
+
+        byte[] StringToByteArray(string input)
```
This could be static if I'm not mistaken

danielmarbach commented on src/NServiceBus.Core.Tests/Serializers/XML/Using_Infer_Type_With_Non_Nested_Class.cs:
```diff
@@ -1,41 +1,57 @@
+            Assert.True(exception.Message.StartsWith("Could not determine type for node:"));
+        }
+
+        byte[] StringToByteArray(string input)
```
This could be static if I'm not mistaken

danielmarbach reviewed (approved):
One nit. Feel free to ignore
